src/v/cluster/shard_placement_table.h

```diff
@@ -0,0 +1,172 @@
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "cluster/types.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/sharded.hh>
+
+#include <absl/container/node_hash_map.h>
+
+namespace cluster {
+
+/// Node-local data structure tracking ntp -> shard mapping for partition
+/// replicas hosted by this node. Both target state and current shard-local
+/// state for each ntp are tracked. Target state is supposed to be modified by
+/// shard_balancer and current shard-local state is supposed to be modified by
+/// controller_backend as it creates/deletes/moves partitions.
+///
+/// Currently shard-local and target states are in-memory and target states
+/// duplicate shard assignments that are stored in topic_table, but in the
+/// future they will be persisted in the kvstore and target states will be set
+/// independently.
+///
+/// Note that in contrast to `cluster::shard_table` (that helps other parts of
+/// the system to find the shard where the `cluster::partition` object for some
+/// ntp resides) this is a more granular table that is internal to the partition
+/// reconciliation process.
+class shard_placement_table
+  : public ss::peering_sharded_service<shard_placement_table> {
+public:
+    // assignment modification methods must be called on this shard
+    static constexpr ss::shard_id assignment_shard_id = 0;
+
+    enum class hosted_status {
+        /// Cross-shard transfer is in progress, we are the destination
+        receiving,
+        /// Normal state, we can start the partition instance.
+        hosted,
+        /// We have transferred our state to somebody else, now our copy must be
+        /// deleted.
+        obsolete,
+    };
+
+    /// Current state of shard-local partition kvstore data on this shard.
+    struct shard_local_state {
+        model::revision_id log_revision;
+        hosted_status status;
+
+        static shard_local_state initial(model::revision_id log_revision) {
+            return shard_local_state{
+              .log_revision = log_revision,
+              .status = hosted_status::hosted,
+            };
+        }
+
+        static shard_local_state receiving(model::revision_id log_revision) {
+            return shard_local_state{
+              .log_revision = log_revision,
+              .status = hosted_status::receiving,
+            };
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream&, const shard_local_state&);
+    };
+
+    /// A struct holding both current shard-local and target states for an ntp.
+    struct placement_state {
+        std::optional<shard_local_state> local;
+        std::optional<shard_placement_target> target;
```

Comment:
would be great to add a comment on what those fields are about the same way as you commented `_next`